### ISSUE TITLE
Client cert authentication fails because Secret Store CSI stores intermediate and root cert in tls.crt, not separately in ca.crt

### ISSUE THREAD

**Issue by andyw248** [Azure/secrets-store-csi-driver-provider-azure#599]
**Have you**

- [X] Read [Troubleshooting Guide](https://azure.github.io/secrets-store-csi-driver-provider-azure/troubleshooting/)
- [X] Searched on [GitHub issues](https://github.com/Azure/secrets-store-csi-driver-provider-azure/issues) and [Discussions](https://github.com/Azure/secrets-store-csi-driver-provider-azure/discussions)

**What steps did you take and what happened:**
[A clear and concise description of what the bug is.]

Bug summary:
Passing a client certificate through nginx-ingress fails when the secret has been generated as a two-part secret {tls.crt and tls.key} by the Secret Store CSI driver. It succeeds when the secret has been manually created as a three-part secret {tls.crt, tls.key, and ca.crt}. 
Can a configuration be added that supports three-part secrets where the intermediate and root cert are stored in a separate ca.crt?

Details:

(1) Deployed a secret provider class for the TLS certificate:

>apiVersion: secrets-store.csi.x-k8s.io/v1alpha1
kind: SecretProviderClass
metadata:
  name: agentdataservice-ingress-tls
spec:
  provider: azure
  secretObjects:                                # secretObjects defines the desired state of synced K8s secret objects
    - secretName: ca-secret
      type: kubernetes.io/tls
      data:
        - objectName: {{ .Values.tlsCert }}
          key: tls.key
        - objectName: {{ .Values.tlsCert }}
          key: tls.crt
  parameters:
    usePodIdentity: "false"
    useVMManagedIdentity: "true"
    userAssignedIdentityID: ""      # If empty it defaults to use the system assigned identity on the VM
    keyvaultName: {{ .Values.keyvaultname }} # the name of the Keyvault
    objects: |
      array:
        - |
          objectName: {{ .Values.tlsCert }}
          objectAlias: tls.key
          objectType: secret                    # Setting the objectType to secret will retrieve the certificate and private key from keyvault.
        - |
          objectName: {{ .Values.tlsCert }}
          objectAlias: tls.crt
          objectType: cert
    tenantId: {{ .Values.tenantid }} # the tenant ID of the KeyVault`

This generates a kubernetes secret that includes the public cert "tls.crt" and the private key "tls.key". The intermediate certificates are concatenated into "tls.crt".

(2) Deployed an nginx ingress manifest with this annotation:

> apiVersion: networking.k8s.io/v1beta1
kind: Ingress
metadata:
  name: agentdataservice-ingress
  namespace: default
  annotations:
    kubernetes.io/ingress.class: nginx
    nginx.ingress.kubernetes.io/ssl-redirect: "true"
    nginx.ingress.kubernetes.io/use-regex: "true"
    nginx.ingress.kubernetes.io/rewrite-target: /$1$2$3
    nginx.ingress.kubernetes.io/auth-tls-verify-client: "optional"
    nginx.ingress.kubernetes.io/auth-tls-secret: "default/ca-secret"
    nginx.ingress.kubernetes.io/auth-tls-verify-depth: "1"
    nginx.ingress.kubernetes.io/auth-tls-pass-certificate-to-upstream: "true"
spec: (...)

Results:

- Https calls are successfully terminated; the secret generated by the Secret Store CSI provider is recognized by Nginx as a TLS certificate.
- The client certificate is not passed through as a header to the app.

**What did you expect to happen:**

According to the kubernetes ingress documentation:

- This document: [creating-certificate-secrets](https://kubernetes.github.io/ingress-nginx/examples/auth/client-certs/#creating-certificate-secrets) explains three ways to create a secret that can be used for client certificate authentication. We have used option (2) from this documentation, which creates a secret with three parts: tls.crt, tls.key, and ca.crt. Our tests show that Nginx passes this certificate through as a header, as expected.
- This document: [client-certificate-authentication](https://kubernetes.github.io/ingress-nginx/user-guide/nginx-configuration/annotations/#client-certificate-authentication) explains the annotations, and how the client certificate will be passed through. This works as documented with a manually uploaded, three-part secret. It does not work with the two-part secret generated by the Secret Store CSI provider: in this case the client certificate is not passed through as a header.

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]

I tinkered with the instructions that you gave in your documentation at: [ingress-tls](https://azure.github.io/secrets-store-csi-driver-provider-azure/configurations/ingress-tls/), by adding the following two sections to my secret provider class:

> - objectName: {{ .Values.tlsCert }}
  key: ca.crt

and:

> - |
  objectName: {{ .Values.tlsCert }}
  objectAlias: ca.crt
  objectType: cert

This can be mounted with the secretProviderClass attribute, but not as a plain secret. Nginx will not recognize it as a TLS secret and will therefore not accept Https requests.

**Which access mode did you use to access the Azure Key Vault instance:**
[e.g. Service Principal, Pod Identity, User Assigned Managed Identity, System Assigned Managed Identity]

System assigned managed identity.

**Environment:**

- Secrets Store CSI Driver version: (use the image tag): 0.0.13
- Azure Key Vault provider version: (use the image tag): 0.0.17
- Kubernetes version: (use `kubectl version` and `kubectl get nodes -o wide`): 1.19.11
- Cluster type: (e.g. AKS, aks-engine, etc): AKS

**Comment by andyw248** [comment]
Update on what got me unblocked:

- Nginx-ingress doesn't really seem to care what's in the ca.crt file within the secret. It just wants that file to exist before it is willing to pass the client certificate to the app.
- As long as the app doesn't evaluate the ca.crt file either, we can just replicate the tls.crt file to stand in and serve as a mock ca.crt. We can do this by using aliases to populate the secretObject.

This leads to the following secret provider class which works as a workaround:

> spec:
  provider: azure
  secretObjects:                                # secretObjects defines the desired state of synced K8s secret objects
    - secretName: ca-secret
      type: Opaque
      data:
        - objectName: tls.key
          key: tls.key
        - objectName: tls.crt
          key: tls.crt
        - objectName: ca.crt
          key: ca.crt
  parameters:
    usePodIdentity: "false"
    useVMManagedIdentity: "true"
    userAssignedIdentityID: ""      # If empty it defaults to use the system assigned identity on the VM
    keyvaultName: {{ .Values.keyvaultname }} # the name of the Keyvault
    objects: |
      array:
        - |
          objectName: {{ .Values.tlsCert }}
          objectAlias: tls.key
          objectType: secret                    # Setting the objectType to secret will retrieve the certificate and private key from keyvault.
        - |
          objectName: {{ .Values.tlsCert }}
          objectAlias: tls.crt
          objectType: cert
        - |
          objectName: {{ .Values.tlsCert }}
          objectAlias: ca.crt
          objectType: key
    tenantId: {{ .Values.tenantid }} # the tenant ID of the KeyVault

Of course this is just a workaround. It doesn't completely solve the problem. The ca.crt file should really contain the intermediate and root certificates; it should not be a duplicate of the tls.crt file. The Secret Store CSI driver should provide a flag that allows storing the intermediate and root certificates in a separate file.

**Comment by aramase** [comment]
@andyw248 Thank you for reporting the issue.

When there is a chain of cert, the provider will get the chain and write to mount as is. We have added an optional feature to construct the chain in the correct order at the provider (server, intermediate, root): https://azure.github.io/secrets-store-csi-driver-provider-azure/configurations/feature-flags/#construct-pem-chain-feature-flag.

Could you try enabling that feature and trying it out? If the 1st certificate in the chain is the server cert, it would work.

**Comment by saifsreubc** [comment]
@aramase , can csi driver extract just ca-certs? Thanks
